src/main/java/io/tiledb/spark/TileDBBatch.java

```diff
@@ -297,14 +381,16 @@ private Pair<List<List<Range>>, Class> buildRangeFromFilter(
         second = nonEmptyDomain.get(f.attribute()).getSecond();
       else second = getMaxValue(f.value().getClass());
       int columnIndex = this.tileDBReadSchema.getColumnId(f.attribute()).get();
-      ranges
-          .get(columnIndex)
-          .add(
-              new Range(
-                  new Pair<>(
-                      addEpsilon(
-                          (Number) f.value(), this.tileDBReadSchema.columnTypes.get(columnIndex)),
-                      second)));
+      Number lowerBound =
+          addEpsilon((Number) f.value(), this.tileDBReadSchema.columnTypes.get(columnIndex));
+      ranges.get(columnIndex).add(new Range(new Pair<>(lowerBound, second)));
+
+      // for qc
+      if (arraySchema.hasAttribute(filterReferences[0])) {
```

Comment:
We repeat this block 4x, maybe we can extract to a function?

Reply:
Working on it.